[FLINK-5906] [table] Add support to register UDAGG in Table and SQL API

shaoxuan-wang opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -911,7 +911,21 @@ class GroupedTable(
+
+    //get the correct expression for UDAGGFunctionCall
+    val input: Seq[Expression] = fieldExprs.zipWithIndex.map {
+      case (Call(name, args), idx) => {
```
This does not catch all cases. The agg function call could be nested in one or more expressions:
```
select('x, weightAvgFun('long, 'int) * 2)
```
So we need to recursively search each expression for UDAGG calls.

It would also be good to make this a utility function in `ProjectionTranslator`.

I think this needs to be added to `Table.select()` as well (non-grouped, non-windowed aggregation).

fhueske replied:
Could also be implemented without `zipWithIndex`:
```
val input: Seq[Expression] = fieldExprs.map {
      case Call(name, args) => {
        val function = table.tableEnv.getFunctionCatalog.lookupFunction(name, args)
        if (function.isInstanceOf[UDAGGFunctionCall]) {
          function
        } else {
          Call(name, args)
        }
      }
      case x => x
    }
```

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala:
```diff
@@ -274,27 +278,73 @@ class CodeGenerator(
-    val accTypes = aggregates.map { a =>
-      a.getClass.getMethod("createAccumulator").getReturnType.getCanonicalName
+    val accTypeClasses = aggregates.map { a =>
+      val accType = TypeExtractor.createTypeInfo(a, classOf[AggregateFunction[_, _]], a.getClass, 1)
```
why doesn't the original approach work anymore?
IMO, it is better to avoid the TypeExtractor if possible, esp. when trying to extract generic types.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala:
```diff
@@ -274,27 +278,73 @@ class CodeGenerator(
+        getUserDefinedMethod(a, "accumulate", Array(accTypeClasses(i)) ++ methodSignaturesList(i))
+          .getOrElse(
+            throw new CodeGenException(
+              s"No matching accumulate method found for aggregate " +
```
add parameter classes to exception, e.g., `... for aggregation function <class name> and parameters <parameter classes>`

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/DStreamUDAGGITCase.scala:
```diff
@@ -0,0 +1,309 @@
+
+    val stream = env
+      .fromCollection(data)
+      .assignTimestampsAndWatermarks(new TimestampAndWatermarkWithOffset(10000))
```
use `.assignAscendingTimestamps(_._1)` instead?

shaoxuan-wang replied:
The reason I used "TimestampAndWatermarkWithOffset(10000)" is that I want to give this test a watermark which can let the late arrival data merges two separate windows.

fhueske replied:
I see, makes sense. Would be good to add a comment about the intention of the test, IMO.

shaoxuan-wang commented:
Thanks for the review, @fhueske , and very good point regarding to the approach to "replace UDAGGFunctionCall". I have addressed your comments. Please take a look.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/DSetUDAGGITCase.scala:
```diff
@@ -0,0 +1,192 @@
+  * programs is possible.
+  */
+@RunWith(classOf[Parameterized])
+class DSetUDAGGITCase(configMode: TableConfigMode)
```
It would be good to implement some unit tests that check the plan and extend `TableTestBase` similar to `UserDefinedTableFunctionTest`.
Some of the tests should validate the equivalence of the Scala expression and String-based API.
Others should validate the generated plan (for Scala Table API and SQL) and check that the validations are correct.
The tests should cover all `select()` methods (`Table`, `GroupedTable`, `WindowGroupedTable`, `OverWindowedTable`).

fhueske replied:
Actually, I think we can just extend the existing aggregation tests and add a user-defined aggregation function to some. This way we can avoid many new tests and just adapt some of the existing ones.

shaoxuan-wang replied:
Sounds good to me. I did not have the UDAGG design across all different aggregation tests, as I feel the current agg tests are a little mess up. It always takes me a while to find the right test cases among all different test files. I put UDAGG test cases into one file which helps me to easily understand what kinds of tests have been covered. I think we need to think about how to reorganize our agg test structure. Considering the short time to freeze feature, let us keep the current structure (I will split the UDAGG into all different agg test files) and massage the tests later.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/DStreamUDAGGITCase.scala:
```diff
@@ -0,0 +1,318 @@
+  * We only test some aggregations until better testing of constructed DataStream
+  * programs is possible.
+  */
+class DStreamUDAGGITCase
```
I think we could extend the existing tests for stream aggregates (group window, over window, non-windowed) and use some UDAGGs there (could be a built-in function which is called as a user-defined agg function).
We should also add the plan tests as described for the batch aggregations.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/expressionDsl.scala:
```diff
@@ -691,6 +692,8 @@ trait ImplicitExpressionConversions {
   implicit def sqlTimestamp2Literal(sqlTimestamp: Timestamp): Expression =
     Literal(sqlTimestamp)
   implicit def array2ArrayConstructor(array: Array[_]): Expression = convertArray(array)
+  implicit def UserDefinedAggFunctionConstructor[T: TypeInformation, ACC](udagg:
```
Methods start with lower case letters.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/expressionDsl.scala:
```diff
@@ -691,6 +692,8 @@ trait ImplicitExpressionConversions {
     Literal(sqlTimestamp)
   implicit def array2ArrayConstructor(array: Array[_]): Expression = convertArray(array)
+  implicit def UserDefinedAggFunctionConstructor[T: TypeInformation, ACC](udagg:
+  AggregateFunction[T, ACC]): UDAGGExpression[T, ACC] = UDAGGExpression(udagg)
```
Do we need this `UDAGGExpression`? Can't we convert directly into `UDAGGFunctionCall`? Imho I would call this `AggregateFunctionCall` and avoid acronyms.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarFunctionCallGen.scala:
```diff
@@ -44,10 +44,10 @@ class ScalarFunctionCallGen(
       .getOrElse(throw new CodeGenException("No matching signature found."))
     val matchingSignature = matchingMethod.getParameterTypes
-    val resultClass = getResultTypeClass(scalarFunction, matchingSignature)
+    val resultClass = getResultTypeClassOfScalaFunction(scalarFunction, matchingSignature)
```
missing `r`

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/aggregations.scala:
```diff
@@ -130,3 +142,63 @@ case class Avg(child: Expression) extends Aggregation {
+    new UDAGGFunctionCall(agg, arg).asInstanceOf[this.type]
+  }
+
+  override def resultType: TypeInformation[_] = TypeExtractor.createTypeInfo(
```
This is again very dangerous. Type extraction fails very often. We should provide a `getResultType` in `AggregateFunction` that the user can implement.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/aggregations.scala:
```diff
@@ -130,3 +142,63 @@ case class Avg(child: Expression) extends Aggregation {
+  override private[flink] def children: Seq[Expression] = args
+
+  // Override makeCopy method in TreeNode, to produce vargars properly
+  override def makeCopy(args: Array[AnyRef]): this.type = {
```
Do we need this? We have no varargs...

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala:
```diff
@@ -274,27 +278,74 @@ class CodeGenerator(
+        }
+
+        if (needMerge) {
+          val methods = checkAndExtractMethods(a, "merge")
```
This check be done as:

```
// get method
val mergeMethod =
            getUserDefinedMethod(a, "merge", Array(accTypeClasses(i), classOf[JIterable[Any]]))
            .getOrElse(
              throw new CodeGenException(
                s"No matching merge method found for aggregate " +
                  s"${a.getClass.getCanonicalName}'.")
            )
          // get class of iterable type
          val iterableTypeClass = mergeMethod.getGenericParameterTypes.apply(1)
            .asInstanceOf[ParameterizedType].getActualTypeArguments.apply(0)
          if (iterableTypeClass != accTypeClasses(i)) {
            throw new CodeGenException(
              s"No matching merge method found for aggregate " +
                s"${a.getClass.getCanonicalName}'.")
          }
```

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala:
```diff
@@ -327,4 +332,56 @@ object ProjectionTranslator {
+    * @param tableEnv the TableEnvironment
+    * @return an expression with correct UDAGGFunctionCall type for UDAGG functions
+    */
+  def replaceUDAGGFunctionCall(field: Expression, tableEnv: TableEnvironment): Expression = {
```
Couldn't we do the replacement during `replaceAggregationsAndProperties`? Save some duplicate code? Actually, `LogicalNode#resolveExpressions` is responsible for looking up calls. Maybe this logic has to be adapted a little bit.

shaoxuan-wang replied:
We will not have the chance to execute LogicalNode#resolveExpressions before get aggNames, projectFields, etc. I actually tried another alternative approach to conduct the replacement in extractAggregationsAndProperties and replaceAggregationsAndProperties (we have to check and handle the UDAGG call carefully in both functions), it works but I do not like that design. It makes the logic of these two methods not completely clean. Also, in over aggregate it will not call extractAggregationsAndProperties and replaceAggregationsAndProperties. So I decide to implement a separate function to handle the UDAGGFunctionCall replacement.

shaoxuan-wang commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala:
```diff
@@ -178,4 +178,24 @@ class BatchTableEnvironment(
+      name: String,
+      f: AggregateFunction[T, ACC])
+  : Unit = {
+    implicit val typeInfo: TypeInformation[T] = TypeExtractor
```
Thanks @twalthr .  If I understand you correctly, you suggest to create a contract method `getResultType` for UDAGG, such that user can provide the result type in case the type extraction fails. Sounds good to me? 
Can you give some examples that when the type extraction will fail (for instance a Row type?) and why it may fail, such that I can add some test cases.

fhueske replied:
The idea would be to always check `getResultType()` first and only use the type extractor if the method is not implemented. So you would not need to enforce a TypeExtractor failure for the tests.

twalthr replied:
Yes, the Row type is a good example for type extraction problems. A user usually doesn't want to use `GenericType<Row>`. Other cases need for custom serializers or types with complex generics.

fhueske replied:
Well, in the cases here, we are not really interested in the TypeInformation itself but the type class. So it would not really matter if you get Row.class from the `GenericTypeInfo` or the `RowTypeInfo`. Nonetheless, I think it is a good idea to have this optional method. It would also be consistent with the other user-defined functions.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/aggregations.scala:
```diff
@@ -130,3 +142,63 @@ case class Avg(child: Expression) extends Aggregation {
+    if (foundSignature.isEmpty) {
+      ValidationFailure(s"Given parameters do not match any signature. \n" +
+                          s"Actual: ${signatureToString(signature)} \n" +
+                          s"Expected: ${signaturesToString(aggregateFunction, "accumulate")}")
```
The signature string includes the Accumulator, which should be removed.

shaoxuan-wang commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/aggregations.scala:
```diff
@@ -130,3 +142,63 @@ case class Avg(child: Expression) extends Aggregation {
+
+  override def toAggCall(name: String)(implicit relBuilder: RelBuilder): AggCall = {
+    val typeFactory = relBuilder.getTypeFactory.asInstanceOf[FlinkTypeFactory]
+    val sqlFunction = AggSqlFunction(name, aggregateFunction, resultType, typeFactory)
```
I was trying to keep this name consistent with `ScalaSqlFunction` and `TableSqlFunction`

shaoxuan-wang commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/AggSqlFunction.scala:
```diff
@@ -0,0 +1,177 @@
+            //do not count accumulator as input
+            var len = inputSig.length
+            if (len > 0 && inputSig(inputSig.length - 1).isArray) {
+              max = 254 // according to JVM spec 4.3.3
```
We want to make sure number of parameters of all codeGened functions within the [min, max], otherwise it will trigger the JVM failure. For accumulate and retract method, the max (for Operand which does not take into account the accumulator) here should be 253, as function pointer itself takes one, and accumulator takes one. Yes, we should consider to add test cases for huge amount of inputs and Array case.

shaoxuan-wang commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/AggSqlFunction.scala:
```diff
@@ -0,0 +1,177 @@
+        for (i <- operandTypes.indices) {
+          if (i < inferredTypes.length - 1) {
+            operandTypes(i) = inferredTypes(i)
+          } else if (null != inferredTypes.last.getComponentType) {
```
Yes, if the last parameter is a component, say Array[Int]. We want to get the type of Int, not the type of Array.

fhueske commented:
Merging
